src/edu/csus/ecs/pc2/clics/API202306/CLICSContestAccess.java

```diff
@@ -0,0 +1,167 @@
+
+    @JsonProperty
+    private String [] capabilities;
+
+    @JsonProperty
+    private CLICSEndpoint [] endpoints;
+
+    IInternalController controller = null;
+
+    /**
+     * Fill in properties for contest state as per 2023-06 spec
+     *
+     * @param sc security info for user making request
+     * @param model
+     * @param controller
+     * @param contestId
+     */
+    public CLICSContestAccess(SecurityContext sc, IInternalContest model, IInternalController controller, String contestId) {
+        this.controller = controller;
+
+        // For each role the connected user has, we enumerate what they can do with each endpoint.
+        ArrayList<String> cap = new ArrayList<String>();
+
+        if(ContestService.isContestStartAllowed(sc)) {
+            cap.add(API_CAPABILITY_CONTEST_START);
+        }
+        if(ContestService.isContestThawAllowed(sc)) {
+            cap.add(API_CAPABILITY_CONTEST_THAW);
+        }
+        if(SubmissionService.isTeamSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_TEAM_SUBMIT);
+        }
+        if(ClarificationService.isTeamSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_TEAM_CLAR);
+        }
+        if(SubmissionService.isProxySubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_SUBMIT);
+        }
+        if(ClarificationService.isProxySubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_CLAR);
+        }
+        if(SubmissionService.isAdminSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_SUBMIT);
+        }
+        if(ClarificationService.isAdminSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_CLAR);
+        }
+        if(AwardService.isAwardsUpdateAllowed(sc)) {
+            cap.add(API_CAPABILITY_AWARDS_UPDATE);
+        }
+        capabilities = cap.toArray(new String[0]);
+
+        /*
+         * As per the spec, here are the possible endpoints.
+         * These should match the class names in this package, eg. Contest -> ContestService, Run -> RunService, etc.
+         */
+        String [] serviceNames = { "Contest", "JudgementType", "Language", "Problem", "Group", "Organization",
+                "Team", "Person", "Account", "State", "Submission", "Judgement", "Run", "Clarification",
+                "Scoreboard", "EventFeed", "Award", "Commentary"
+        };
+
+        ArrayList<CLICSEndpoint> epList = new ArrayList<CLICSEndpoint>();
+
+        for (String service: serviceNames) {
+            if(!invokeServiceEndpointAccessMethod(epList, service, sc)) {
+                if(Utilities.isDebugMode()) {
+                    System.err.println("No endpoint access for service " + service);
+                }
+                controller.getLog().log(Log.WARNING, "No endpoint access for service " + service);
+            }
+        }
+        if(!epList.isEmpty()) {
+            endpoints = epList.toArray(new CLICSEndpoint[0]);
```

Comment:
The spec says that the `endpoints` array "may be empty".  However, putting the above assignment inside the statement `if (!epList.isEmpty()) {...` will result in a NULL `endpoints` value (not an "empty array") if the `epList` is indeed empty.  I think maybe the `if` statement should be removed, leaving the statement `endpoints = epList.toArray(new CLICSEndpoint[0]);` to be executed unconditionally?  This would make it function similarly to the creation of the `capabilities` property above.   Or am I missing something?

Reply:
I agree.  That makes sense.  Since endpoints is required by the spec, it should not be null.  Next push will fix that.

Reply:
Interesting.  Personally, I find `if(!eplist.isEmpty())` easier to grok than `if(eplist.isEmpty() == false) `.  The former reads to me like `is not empty` (which is ultimately the intended test), whereas with the latter I find I have to mentally explicitly ask "wait; if it's `isEmpty() is false`  that means that the `isEmpty()` function is returning `false` which means the condition is that it's not empty which means the `if` condition is actually true; but wait; did I do that logic correctly?"    Like I said, the former reads cleanly to me whereas with the latter I literally find myself doing the above mental gymnastics every time.   But to each his own :)